include/hal/dts-hal.sh

```diff
@@ -0,0 +1,206 @@
+# function for a tool but we do not care about its output.
+#
+# This tool wrapper should only be used with tools which communicate with
+# hardware or firmware (read or write, etc.).
+#
+# TODO: this wrapper deals with arguments as well as with stdout, stderr, and $?
+# redirection, but it does not read and redirect stdin (this is not used in any
+# mocking functions or tools right now).
+  # Gets toolname, e.g. poweroff, dmidecode. etc.:
+  local _tool="$1"
+  # Gets mocking function name:
+  local _mock_func="$2"
+  # It checks if _mock_func contains smth with _mock at the end, if not -
+  # mocking function is not provided and some common mocking func. will be used
+  # instead:
+  if ! echo "$_mock_func" | grep "_mock" &> /dev/null; then
+    unset _mock_func
+    shift 1
+  else
+    shift 2
+  fi
+  # Other arguments for this function are the arguments which are sent to a tool
+  # e.g. -s system-vendor for dmidecode, etc.:
+  local _arguments=( "$@" )
+
+  if [ -n "$DTS_TESTING" ]; then
+    # This is the order of calling mocking functions:
+    # 1) FUNCTIONNAME_mock;
+    # 2) TOOLNAME_FUNCTIONNAME_mock;
+    # 3) TOOLNAME_common_mock;
+    # 4) common_mock - last resort.
+    if [ -n "$_mock_func" ] && type $_mock_func &> /dev/null; then
+      $_mock_func "${_arguments[@]}"
+    elif type ${_tool}_${_mock_func} &> /dev/null; then
+      ${_tool}_${_mock_func} "${_arguments[@]}"
+    elif type ${_tool}_common_mock &> /dev/null; then
+      ${_tool}_common_mock "${_arguments[@]}"
+    else
+      common_mock $_tool
+    fi
+
+    return $?
+  fi
+
+  # If not testing - call tool with the arguments instead:
+  $_tool "${_arguments[@]}"
+
+  return $?
+}
+
+################################################################################
+# Other funcs.
+################################################################################
+check_for_opensource_firmware()
+{
+  echo "Checking for Open Source Embedded Controller firmware..."
+  $DASHARO_ECTOOL check_for_opensource_firm_mock info > /dev/null 2>&1
+
+  return $?
+}
+
+fsread_tool(){
+# This func is an abstraction for proper handling of fs hardware-specific (e.g.
+# checking devtmpfs, or sysfs, or some other fs that changes its state due to
+# changes in hardware and/or firmware) reads by tool_wrapper.
+#
+# This function does not have arguments in common understanding, it takes a
+# command, that is reading smth from some fs, and its arguments as an only
+# argument. E.g. if you want to check tty1 device presence:
+#
+# fsread_tool test -f /dev/tty1
+  local _command="$1"
+
+  $_command "$@"
+
+  return $?
+}
+
+cap_upd_tool(){
+# This func is an abstraction for proper handling of UEFI Capsule Update driver
+# writing by the tool_wrapper. arguments: capsule update file path, e.g.:
+#
+# capsule_update_tool /tmp/firm.cap
+  local _capsule="$1"
+  local _result
+
+  cat "$_capsule" > "$CAP_UPD_DEVICE"
+  _result="$?"
+
+  if [ "$?" != "0" ]; then
```

Comment:
```suggestion
  if [ "$_result" != "0" ]; then
```

Earlier you use `_result="$?"` which will always result in `$?` changing to `0`

Reply:
Fixed [here](https://github.com/Dasharo/dts-scripts/compare/ed6c6089caa0e541f15150b15a79a7e26c6ae365..9aa00a04166f712f638acc875e4004c28ece5c32)